business/openshift_oauth.go

```diff
@@ -0,0 +1,194 @@
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+
+	kube "k8s.io/client-go/kubernetes"
+)
+
+type OpenshiftOAuthService struct {
+	k8s kubernetes.IstioClientInterface
+}
+
+type OAuthMetadata struct {
+	AuthorizationEndpoint string `json:"authorizationEndpoint"`
+}
+
+// Structure that's returned by the openshift oauth authorization server.
+// It defaults to following the snake_case format, so we parse it to something
+// more usable on our side.
+type OAuthAuthorizationServer struct {
+	AuthorizationEndpoint string `json:"authorization_endpoint"`
+}
+
+type OAuthUser struct {
+	Metadata OAuthUserMetadata `json:"metadata"`
+}
+
+type OAuthUserMetadata struct {
+	Name string `json:"name"`
+}
+
+type OAuthRoute struct {
+	Spec OAuthRouteSpec `json:"spec"`
+}
+
+type OAuthRouteSpec struct {
+	Host string             `json:"host"`
+	TLS  *OAuthRouteTLSSpec `json:"tls,omitempty"`
+}
+
+type OAuthRouteTLSSpec struct {
+	Termination string `json:"termination"`
+}
+
+const serverPrefix = "https://openshift.default.svc/"
+
+func (in *OpenshiftOAuthService) Metadata() (metadata *OAuthMetadata, err error) {
+	var server *OAuthAuthorizationServer
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	response, err := request("GET", ".well-known/oauth-authorization-server", nil)
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	err = json.Unmarshal(response, &server)
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	redirectURL, err := getKialiRoutePath()
+
+	metadata = &OAuthMetadata{}
+
+	metadata.AuthorizationEndpoint = fmt.Sprintf("%s?client_id=%s&redirect_uri=%s&response_type=%s", server.AuthorizationEndpoint, "kiali", url.QueryEscape(*redirectURL), "token")
+
+	return metadata, nil
+}
+
+func (in *OpenshiftOAuthService) ValidateToken(token string) error {
+	k8sConfig, err := kubernetes.ConfigClient()
+
+	if err != nil {
+		return fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	k8sConfig.BearerToken = token
+
+	k8s, err := kube.NewForConfig(k8sConfig)
+
+	if err != nil {
+		return fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	_, err = k8s.Discovery().ServerVersion()
+
+	if err != nil {
+		return fmt.Errorf("could not get info from Openshift: %v", err)
+	}
+
+	return nil
+}
+
+func (in *OpenshiftOAuthService) GetUserInfo(token string) (*OAuthUser, error) {
+	var user *OAuthUser
+
+	response, err := request("GET", "oapi/v1/users/~", &token)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	err = json.Unmarshal(response, &user)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
```

Comment:
this error message is inaccurate. This doesn't deal with connecting with OpenShift. It has to do with parsing the results. Error message should be something like "Could not parse oauth details from OpenShift".